monitoring/rid_qualifier/host/routes.py

```diff
@@ -1,69 +1,148 @@
 import flask
 import os
+import pathlib
+import requests
 
 from . import config
 from . import forms
 from . import tasks
 
-from flask import render_template, request, make_response, redirect, url_for
+from google.oauth2 import id_token
+from google_auth_oauthlib.flow import Flow
+import google.auth.transport.requests
+from flask import render_template, request, make_response, redirect, url_for, session
+from functools import wraps
+from pip._vendor import cachecontrol
 from werkzeug.exceptions import HTTPException
 from werkzeug.utils import secure_filename
 
 from monitoring.monitorlib import versioning, auth_validation
 from monitoring.rid_qualifier.host import webapp
 
 
+client_secrets_file = os.path.join(
+    pathlib.Path(__file__).parent,
+    'client_secret.json')
+
+flow = Flow.from_client_secrets_file(
+    client_secrets_file=client_secrets_file,
+    scopes=[
+        'https://www.googleapis.com/auth/userinfo.profile',
+        'https://www.googleapis.com/auth/userinfo.email',
+        'openid'],
+    redirect_uri=f'{config.Config.RID_HOST_URL}/callback')
+
+
+def login_required(function):
+    @wraps(function)
+    def decorated_function(*args, **kwargs):
+        if 'google_id' not in session:
+            return redirect(url_for('login', next=request.url))
+        return function(*args, **kwargs)
+    return decorated_function
+
+
+@webapp.route('/login')
+def login():
+    # make sure session is empty
+    session.clear()
+    authorization_url, state = flow.authorization_url()
+    session['state'] = state
+    return redirect(authorization_url)
+
+
+@webapp.route('/callback')
+def callback():
+    flow.fetch_token(authorization_response=request.url)
+
+    credentials = flow.credentials
+    request_session = requests.session()
+    cached_session = cachecontrol.CacheControl(request_session)
+    token_request = google.auth.transport.requests.Request(
+        session=cached_session)
+
+    id_info = id_token.verify_oauth2_token(
+        id_token=credentials._id_token,
+        request=token_request,
+        audience=config.Config.GOOGLE_CLIENT_ID
+    )
+
+    session['google_id'] = id_info.get('sub')
+    session['name'] = id_info.get('name')
+    return redirect('/')
+
+
+@webapp.route('/logout')
+def logout():
+    session.clear()
+    return render_template('logout.html')
+
+
 @webapp.route('/')
+@login_required
 def home_page():
-  return render_template('home.html', title='Home', greetings='Hello RID Host !!')
+    if session.get('google_id'):
```

Comment:
Can we remove this check for `google_id` since this function is decorated with `@login_required`?